include/bm/bm_sim/field_lists.h

```diff
@@ -82,24 +102,31 @@ class FieldList {
   //! Returns true if the FieldList contains the given field, identified by the
   //! header id and the offset of the field in the header
   bool contains(header_id_t header, int field_offset) const {
-    auto it = fields_set.find({header, field_offset});
+    field_t f = {header, field_offset};
+    auto it = fields_set.find(field_list_member_t(f));
     return it != fields_set.end();
   }
 
  private:
   struct FieldKeyHash {
-    std::size_t operator()(const field_t& f) const {
+    std::size_t operator()(const field_list_member_t& flm) const {
       std::size_t seed = 0;
-      boost::hash_combine(seed, f.header);
-      boost::hash_combine(seed, f.offset);
-
+      if (flm.type() == typeid(field_t)) {
+        field_t f = boost::get<field_t>(flm);
+        boost::hash_combine(seed, f.header);
+        boost::hash_combine(seed, f.offset);
+      } else if (flm.type() == typeid(constant_t)) {
+        constant_t c = boost::get<constant_t>(flm);
+        boost::hash_combine(seed, c.value);
+        boost::hash_combine(seed, c.nbits);
+      }
       return seed;
     }
   };
 
  private:
-  std::vector<field_t> fields{};
-  std::unordered_set<field_t, FieldKeyHash> fields_set{};
+  std::vector<field_list_member_t> fields{};
+  std::unordered_set<field_list_member_t, FieldKeyHash> fields_set{};
```

Comment:
You can just stick with fields in this set (`field_t`) and maybe rename the `contains` method to `contains_field`